pkg/kube/inject/app_probe.go

```diff
@@ -77,6 +86,18 @@ func convertAppProber(probe *corev1.Probe, newURL string, statusPort int) *corev
 	return p
 }
 
+// convertAppProberTCPSocket returns an overwritten `Probe` (TcpSocket) for pilot agent to take over.
+func convertAppProberTCPSocket(probe *corev1.Probe, newURL string, statusPort int) *corev1.Probe {
+	p := probe.DeepCopy()
+	// the sidecar intercepts all tcp connections, so we change it to a special HTTP probe
+	p.HTTPGet = &corev1.HTTPGetAction{}
+	p.HTTPGet.Port = intstr.FromInt(statusPort)
+	// the path contains the original port of the application
```

Comment:
same comments as john. we should not rely on client indicating the port, instead the original tcp port value can be marshaled to json env var and pilot agent only allows that port.